mayastor/src/grpc/mayastor_grpc.rs

```diff
@@ -32,36 +33,75 @@ use crate::{
     nexus_uri::NexusBdevError,
     subsys::PoolConfig,
 };
+use futures::FutureExt;
 use nix::errno::Errno;
 use rpc::mayastor::*;
-use std::{convert::TryFrom, ops::Deref, time::Duration};
+use std::{convert::TryFrom, fmt::Debug, ops::Deref, time::Duration};
 use tonic::{Request, Response, Status};
 #[derive(Debug)]
 struct UnixStream(tokio::net::UnixStream);
 
+use ::function_name::named;
+use std::panic::AssertUnwindSafe;
+
+impl GrpcClientContext {
+    #[track_caller]
+    pub fn new<T>(req: &Request<T>, fid: &str) -> Self
+    where
+        T: Debug,
+    {
+        Self {
+            args: format!("{:?}", req.get_ref()),
+            id: fid.to_string(),
+        }
+    }
+}
+
 #[derive(Debug)]
 pub struct MayastorSvc {
     name: String,
     interval: Duration,
-    rw_lock: tokio::sync::RwLock<bool>,
+    rw_lock: tokio::sync::RwLock<Option<GrpcClientContext>>,
 }
 
 #[async_trait::async_trait]
-impl<F, R> Serializer<F, R> for MayastorSvc
+impl<F, T> Serializer<F, T> for MayastorSvc
 where
-    F: core::future::Future<Output = R> + Send + 'static,
-    R: Send,
+    T: Send + 'static,
+    F: core::future::Future<Output = Result<T, Status>> + Send + 'static,
 {
-    async fn locked(&self, f: F) -> R {
-        let mut lock = self.rw_lock.write().await;
-        trace!(?self.name, "locked");
-        assert!(!*lock);
-        *lock = true;
-        let output = f.await;
-        tokio::time::sleep(self.interval).await;
-        trace!(?self.name, "unlocked");
-        *lock = false;
-        output
+    async fn locked(&self, ctx: GrpcClientContext, f: F) -> Result<T, Status> {
+        let mut guard = self.rw_lock.write().await;
+
+        // Store context as a marker of to detect abnormal termination of the
+        // request. Even though AssertUnwindSafe() allows us to
+        // intercept asserts in underlying method strategies, such a
+        // situation can still happen when the high-level future that
+        // represents gRPC call at the highest level (i.e. the one created
+        // by gRPC server) gets cancelled (due to timeout or somehow else).
+        // This can't be properly intercepted by 'locked' function itself in the
+        // first place, so the state needs to be cleaned up properly
+        // upon subsequent gRPC calls.
+        if let Some(c) = guard.replace(ctx) {
+            warn!("{}: gRPC method timed out, args: {}", c.id, c.args);
+        }
+
+        let fut = AssertUnwindSafe(f).catch_unwind();
```

Comment:
hmm so we allow the gRPC to be retried because the tonic future but the actual mayastor future sent with "rpc_submit" is still running, so we loose the "serialization" that this code was adding to begin with.
Is there anything that can be done about that?

Reply:
This situation is two-fold. On one hand we have a running rpc_submit-related future, which might result in a de-serialised invocation as you are pointing out. So it would be nice to make sure this future has completed before proceeding. And this is doable, IMHO (we could introduce one-shot channel and make the receiver part be part of GRPC context, so we can complete the call the next time we see the existing context). But then another question arises: for how long should we wait to not block forever, as we want to avoid a situation when one blocking call blocks all other methods.
We definitely can improve our gRPC serialisation (use fine-grained locks instead of one global lock, etc), but I would do it on behalf of another changeset, where we can also handle this 'serialisation escape' question.

Reply:
Yeah, I'm not sure how we'd want to handle the case a call is taking forever, though we could argue it can still happen anyway if the gRPC client does not set a timeout.
But yeah, not sure what we'd want to do... @gila?